velox/functions/lib/aggregates/AverageAggregateBase.h

```diff
@@ -0,0 +1,423 @@
+#include "velox/expression/FunctionSignature.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/DecimalAggregate.h"
+#include "velox/type/DecimalUtil.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/DecodedVector.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::functions::aggregate {
+
+namespace {
+
+/// Translate selected rows of decoded to the corresponding rows of its base
+/// vector.
+SelectivityVector translateToInnerRows(
+    const DecodedVector& decoded,
+    const SelectivityVector& rows) {
+  VELOX_DCHECK(!decoded.isIdentityMapping());
+  if (decoded.isConstantMapping()) {
+    auto constantIndex = decoded.index(rows.begin());
+    SelectivityVector baseRows{constantIndex + 1, false};
+    baseRows.setValid(constantIndex, true);
+    baseRows.updateBounds();
+    return baseRows;
+  } else {
+    SelectivityVector baseRows{decoded.base()->size(), false};
+    rows.applyToSelected(
+        [&](auto row) { baseRows.setValid(decoded.index(row), true); });
+    baseRows.updateBounds();
+    return baseRows;
+  }
+}
+
+/// Return the selected rows of the base vector of decoded corresponding to
+/// rows. If decoded is not identify mapping, baseRowsHolder contains the
+/// selected base rows. Otherwise, baseRowsHolder is unset.
+const SelectivityVector* getBaseRows(
+    const DecodedVector& decoded,
+    const SelectivityVector& rows,
+    SelectivityVector& baseRowsHolder) {
+  const SelectivityVector* baseRows = &rows;
+  if (!decoded.isIdentityMapping() && rows.hasSelections()) {
+    baseRowsHolder = translateToInnerRows(decoded, rows);
+    baseRows = &baseRowsHolder;
+  }
+  return baseRows;
+}
+
+template <typename TSum>
+struct SumCount {
+  TSum sum{0};
+  int64_t count{0};
+};
+
+} // namespace
+
+/// Partial aggregation produces a pair of sum and count.
+/// Count is BIGINT() while sum  and the final aggregates type depends on
+/// the input types:
+///       INPUT TYPE    |     SUM             |     AVG
+///   ------------------|---------------------|------------------
+///     REAL            |     DOUBLE          |    REAL
+///     ALL INTs        |     DOUBLE          |    DOUBLE
+///
+template <typename TInput, typename TAccumulator, typename TResult>
+class AverageAggregateBase : public exec::Aggregate {
+ public:
+  explicit AverageAggregateBase(TypePtr resultType)
+      : exec::Aggregate(resultType) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(SumCount<TAccumulator>);
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    setAllNulls(groups, indices);
+    for (auto i : indices) {
+      new (groups[i] + offset_) SumCount<TAccumulator>();
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValuesImpl(groups, numGroups, result);
```

Comment:
This function seems to be called from only one place: extractValues(). If that's the case, then this function is not needed and its logic can be folded into the caller. Spark version can simply override extractValues(). Let me know if I miss something.

Reply:
Yes, that's the case. Removed extractValuesImpl and moved its logic to extractValues.